### PR TITLE
Prepare monitoring solutions in Keep Test project

### PR DESCRIPTION
**Pull request by electropolis**
Refs #3012 
Refs #3279

 - added deployment for prometheus
 - added deployment for grafana
 - added storageclass to insert persistent storage using cloud provider
 - prepared PersistentVolumeClaim both for prometheus and grafana
 - prepared basic service using NodePort to use with Ingress
 - preaparing README file with instructions

### CODE REVIEW

**Comment by electropolis** [comment]
Please don't merge it yet as it's still in progress

File: infrastructure/kube/keep-test/monitoring/README.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    --clusterrole cluster-admin \
+    --user $ACCOUNT
+```
+
+*! This role can be created by Owner in GCP IAM*
+
+3. Create `clusterRole`
+
+  ```bash
+  kubectl create -f clusterRole.yaml
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/README.md:38]:
```suggestion
  kubectl create -f clusterrole.yaml
```

File: infrastructure/kube/keep-test/monitoring/README.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  ```
+
+#### II. Create `ConfigMap` to put configuration externally
+
+1. Execute `create` command to create configmap
+    
+  ```bash
+  kubectl create -f config-map.yaml
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/README.md:46]:
```suggestion
  kubectl create -f configmap.yaml
```

File: infrastructure/kube/keep-test/monitoring/configmap.yaml [hunk 3]
```diff
@@ -0,0 +1,656 @@
+---
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/configmap.yaml:1]: Maybe we could split this file into: `prometheus-configmap.yaml` and `grafana-configmap.yaml`?

File: infrastructure/kube/keep-test/monitoring/configmap.yaml [hunk 4]
```diff
@@ -0,0 +1,656 @@
...
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  namespace: monitoring
+  name: prometheus-conf
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/configmap.yaml:6]: Do we need `-conf` suffix?
```suggestion
  name: prometheus
```

File: infrastructure/kube/keep-test/monitoring/configmap.yaml [hunk 5]
```diff
@@ -0,0 +1,656 @@
...
+  labels:
+    name: prometheus-conf
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/configmap.yaml:8]: Do we need the labels?

```suggestion
```

File: infrastructure/kube/keep-test/monitoring/grafana-deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+      name: grafana
+      labels:
+        app: grafana
+    spec:
+      securityContext:
+          runAsUser: 472
+          fsGroup: 472
+      containers:
+        - name: grafana
+          image: grafana/grafana:9.0.5
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/grafana-deployment.yaml:23]: Can we try updating to `9.1.2`? This is the latest published.

File: infrastructure/kube/keep-test/monitoring/grafana-deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+              memory: 500M
+              cpu: "500m"
+          volumeMounts:
+            - name: grafana-storage
+              mountPath: /var/lib/grafana
+            - name: grafana-dashboards-files
+              mountPath: /var/lib/grafana/dashboards
+            - name: grafana-datasources
+              mountPath: /etc/grafana/provisioning/datasources
+              readOnly: false
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/grafana-deployment.yaml:41]: Do we need `readOnly` entries? They are set to `false` by default anyway.

File: infrastructure/kube/keep-test/monitoring/configmap.yaml [hunk 8]
```diff
@@ -0,0 +1,656 @@
...
+          action: replace
+          target_label: kubernetes_name
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: grafana
+  namespace: monitoring
+data:
+  prometheus.yaml: |-
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/configmap.yaml:194]:
```suggestion
  datasources.yaml: |-
```

File: infrastructure/kube/keep-test/monitoring/grafana-pvc.yaml [hunk 9]
```diff
@@ -0,0 +1,14 @@
+---
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: grafana-pvc
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/grafana-pvc.yaml:5]:
```suggestion
  name: grafana
  namespace: monitoring
```

File: infrastructure/kube/keep-test/monitoring/grafana-service.yaml [hunk 10]
```diff
@@ -0,0 +1,17 @@
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: grafana
+  namespace: monitoring
+  annotations:
+      prometheus.io/scrape: 'true'
+      prometheus.io/port: '3000'
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/grafana-service.yaml:9]: How are these annotations used?

File: infrastructure/kube/keep-test/monitoring/grafana-service.yaml [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+spec:
+  selector:
+    app: grafana
+  type: NodePort
+  ports:
+    - port: 3000
+      targetPort: 3000
+      nodePort: 32000
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/grafana-service.yaml:17]: Do we need to set a fixed `nodePort`?

File: infrastructure/kube/keep-test/monitoring/grafana-service.yaml [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+  annotations:
+      prometheus.io/scrape: 'true'
+      prometheus.io/port: '3000'
+spec:
+  selector:
+    app: grafana
+  type: NodePort
+  ports:
+    - port: 3000
+      targetPort: 3000
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/grafana-service.yaml:16]: We named the port in the deployment 
```yaml
          ports:
            - name: grafana
              containerPort: 3000
```
so maybe we could use the name here?
```suggestion
      targetPort: grafana
```

File: infrastructure/kube/keep-test/monitoring/monitoring-ingress.yaml [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+apiVersion: networking.k8s.io/v1
+kind: Ingress
+metadata:
+  name: monitoring
+  namespace: monitoring
+  annotations:
+    kubernetes.io/ingress.class: nginx
+spec:
+  rules:
+    # Use the host you used in your kubernetes Ingress Configurations
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/monitoring-ingress.yaml:13]: Is this comment needed?

```suggestion
```

File: infrastructure/kube/keep-test/monitoring/monitoring-ingress.yaml [hunk 14]
```diff
@@ -0,0 +1,44 @@
+---
+## Nginx Ingress
+## Follow https://devopscube.com/setup-ingress-kubernetes-nginx-controller/
+apiVersion: networking.k8s.io/v1
+kind: Ingress
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/monitoring-ingress.yaml:5]: Is ingress enough to expose the service publicly or is some additional GCP configuration needed?

File: infrastructure/kube/keep-test/monitoring/prometheus-pvc.yaml [hunk 15]
```diff
@@ -0,0 +1,14 @@
+---
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: prometheus-pvc
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/prometheus-pvc.yaml:5]: Could we drop `-pvc` suffix?

```suggestion
  name: prometheus
```

File: infrastructure/kube/keep-test/monitoring/prometheus-service.yaml [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+    prometheus.io/scrape: 'true'
+    prometheus.io/port: '9090'
+spec:
+  selector:
+    app: prometheus
+  type: NodePort
+  ports:
+    - port: 9090
+      targetPort: 9090
+      nodePort: 30000
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/prometheus-service.yaml:17]: Do we need a fixed `nodePort`?

Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/prometheus-service.yaml:17]: Can we use `prometheus` as we named the port in the deployment?

```suggestion
      nodePort: prometheus
```

File: infrastructure/kube/keep-test/monitoring/README.md [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+
+The Kubernetes Prometheus monitoring stack has the following components.
+
+1. Prometheus Server
+2. Alert Manager
+3. Grafana
+
+![Schema](https://devopscube.com/wp-content/uploads/2022/01/kubernetes-1024x498.png)
+
+### <u>_Prometheus_</u>
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/README.md:14]: Do we need `<u>` tags?
```suggestion
### Prometheus
```
or 

```suggestion
### _Prometheus_
```

File: infrastructure/kube/keep-test/monitoring/README.md [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+
+#### I. Create Namespace & ClusterRole
+
+1. Execute the following command to create a new namespace named *monitoring*.
+  
+  ```bash
+  kubectl create namespace monitoring
+  ```
+
+2. Additionaly Cluster Role Binding is required for prometheus  to view k8s metrics. In GCP a `clusterrolebinding` is required to be created.
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/README.md:26]:
```suggestion
2. Additional Cluster Role Binding is required for prometheus to view k8s metrics. In GCP a `clusterrolebinding` is required to be created.
```

File: infrastructure/kube/keep-test/monitoring/README.md [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+
+#### II. Create `ConfigMap` to put configuration externally
+
+1. Execute `create` command to create configmap
+    
+  ```bash
+  kubectl create -f configmap.yaml
+  ```
+
+>Some words to that `ConfigMap`
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/README.md:51]: Where does this quote come from? Could we add a reference?

File: infrastructure/kube/keep-test/monitoring/README.md [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+
+This object will use `NodePort` type for accessing the Prometheus service by using additional VPN service that is routed to kubernetes network. 
+
+  ```bash
+  kubectl create -f prometheus-service.yaml
+  ```
+
+Service should be available under
+
+http://prometheus.monitoring.svc.cluster.local:9090
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/README.md:88]: It is! :tada:

File: infrastructure/kube/keep-test/monitoring/README.md [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+
+  ```bash
+  kubectl create -f grafana-deployment.yaml
+  ```
+
+#### V. Create Grafana Service
+
+  ```bash
+  kubectl create -f grafana-service.yaml
+  ```
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/README.md:124]: Can we also add a link othe grafana instance?

File: infrastructure/kube/keep-test/monitoring/grafana-configmap.yaml [hunk 22]
```diff
@@ -0,0 +1,477 @@
...
+          {
+            "allValue": null,
+            "current": {
+              "selected": true,
+              "text": "keep-client-0.default.svc.cluster.local:9601",
+              "value": "keep-client-0.default.svc.cluster.local:9601",
+              "text": "keep-client-1.default.svc.cluster.local:9601",
+              "value": "keep-client-1.default.svc.cluster.local:9601",
+              "text": "bst-a01.test.keep.boar.network:9601:9601",
+              "value": "bst-a01.test.keep.boar.network:9601:9601"
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/grafana-configmap.yaml:433]: Are the duplicated ports intentional?

```suggestion
              "text": "bst-a01.test.keep.boar.network:9601",
              "value": "bst-a01.test.keep.boar.network:9601"
```

File: infrastructure/kube/keep-test/monitoring/grafana-configmap.yaml [hunk 23]
```diff
@@ -0,0 +1,477 @@
...
+          "5m",
+          "15m",
+          "30m",
+          "1h",
+          "2h",
+          "1d"
+        ]
+      },
+      "timezone": "utc",
+      "title": "clients-dashboard",
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/grafana-configmap.yaml:474]:
```suggestion
      "title": "Bootstrap Nodes",
```

File: infrastructure/kube/keep-test/monitoring/grafana-configmap.yaml [hunk 24]
```diff
@@ -0,0 +1,477 @@
...
+          "editable": false,
+          "updateIntervalSeconds": 10,
+          "allowUiUpdates": false,
+          "options": {
+            "path": "/var/lib/grafana/dashboards"
+          }
+        }
+      ]
+    }
+  clients-dashboard.json: |-
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/grafana-configmap.yaml:43]:
```suggestion
  bootstrap-dashboard.json: |-
```

File: infrastructure/kube/keep-test/monitoring/kustomization.yaml [hunk 25]
```diff
@@ -0,0 +1,15 @@
+---
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/kustomization.yaml:1]: How is `kustomization.yaml` used? Can we add some info to the documentation?

File: infrastructure/kube/keep-test/monitoring/monitoring-ingress.yaml [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
...OUJIRVlPS3oxRFZxNWxHRFhoNFMyTStpOU1pck5nUlcvL0NFRGhRUVVMZmtBTDgxMERPQmxsMXRRRUpGK3cKb1V6dWt6U1lkK1hTSnhicTM5YTF1ZGJldTNZU1ljeC8wTEEweGFQOW1sN1l1NXUraUZ4NGhwcnYyL2UrVklZQQpzTWV4WkZSODA3Q3M5YXN5MkdFT1l2aEdKb0VDZ1lFQTUzVm1weFlQbDFOYTVTMElJbEpuYm40dTl0RHpwYm5TCnpsMjBVQ3Q0d0N4STR6YjY1S1o4V1VaYlFzVTVaZ0VqTmxJWURXUisrd3kwVXh2SmNxUG5nS0xuOEdoSzhvOVEKeVJuR2dSYXAxWmNuUEdsbGdCeHQzM0s5TDNWMmJzMXBPcGJKMGlpOVdySWM4MU1wUVFpQjZ1RDRSZ216M0ZWSQpnUk5Ec2ZHS0xyRUNnWUVBbWY5ZXRqc3RUbGJHZVJ2dDVyUlB4bmR0dFNvTysyZ1RXWnVtSmM0aG1RMldYOWFWCjlKNFZTMWJqa1RrWHV5d0NGMis0dlNmeWxaZFd6U1M3bmMyOFV3dnNmekxYZjVxV05tV3hIYnBTdFcwVnp3c1QKeENyVWFDczd2ODlWdXZEMTVMc1BKZ0NWT0FSalVjd0FMM0d2aDJNeVd4ZE9pQ0g5VFRYd0lJYjFYQUVDZ1lBMwp4ZUptZ0xwaERJVHFsRjlSWmVubWhpRnErQTY5OEhrTG9TakI2TGZBRnV1NVZKWkFZcDIwSlcvNE51NE4xbGhWCnpwSmRKOG94Vkc1ZldHTENiUnhyc3RXUTZKQ213a0lGTTJEUjJsUXlVNm53dExUd21la2YzdFlYaVlad1RLNysKbnpjaW5RNkR2RWVkbW54bVgxWnU4cWJndVpYTmtmOVdtdjNFOHg4SkFRS0JnUUNNeDFWNHJIcUpwVXJMdkRVVQo4RzhXVGNrT2VFM2o2anhlcHMwcnExdEd1cE9XWW5saFlNYyt5VkMzMDZUc2dXUmJ5R1Y4YWNaRkF4WS9Ub2N5CmxpcXlUS1NGNUloYXhZQVpRTzVkOU1oTmN0bTRReDNaOUtTekZ5ZG01QlZVL0grMFFmUnRwM29TeFVneXRZNXkKV3ZDTFZ5bmNGZlZpL0VkaTdaZHM2aW82QVE9PQotLS0tLUVORCBQUklWQVRFIEtFWS0tLS0tCg==
```
Comment (nkuba) [inline infrastructure/kube/keep-test/monitoring/monitoring-ingress.yaml:44]: Why is this section commented out? Can we at least add a clarification inline with a `# TODO:` perhaps?

**Review by nkuba — COMMENTED** [review]
There are still some outstanding comments from the previous review round.

File: infrastructure/kube/keep-test/monitoring/README.md (hunk 17 above)
Comment (electropolis) [reply infrastructure/kube/keep-test/monitoring/README.md:14]: It's for underscore to split that README into two sections

File: infrastructure/kube/keep-test/monitoring/README.md (hunk 19 above)
Comment (electropolis) [reply infrastructure/kube/keep-test/monitoring/README.md:51]: This makes a quote for the text below to make some kind of annotations

File: infrastructure/kube/keep-test/monitoring/monitoring-ingress.yaml [hunk 27]
```diff
@@ -0,0 +1,44 @@
+---
```
Comment (pdyraga) [inline infrastructure/kube/keep-test/monitoring/monitoring-ingress.yaml:1]: I don't see this file referenced anywhere. Should we add it to `README.md`?

File: infrastructure/kube/keep-test/monitoring/README.md [hunk 28]
```diff
@@ -0,0 +1,162 @@
...
+  namespace: monitoring
+```
+
+
+### Grafana
+
+#### I. Create Grafana ConfigMap
+
+   ```bash
+   kubectl create -f grafana-configmap.yaml
```
Comment (pdyraga) [inline infrastructure/kube/keep-test/monitoring/README.md:112]: Should we have at least a template of this file somewhere?

File: infrastructure/kube/keep-test/monitoring/README.md [hunk 29]
```diff
@@ -0,0 +1,162 @@
...
+   ```
+
+Notice the **url** key in the ConfigMap file that refers to prometheus FQDN that comes from combination of `metadata` in the prometheus-service YAML file.
+
+#### II. Create Grafana PersistentVolumeClaim
+
+  ```bash
+  kubectl create -f grafana-pvc.yaml
```
Comment (pdyraga) [inline infrastructure/kube/keep-test/monitoring/README.md:120]: Same question as for Prometheus PersistentVolumeClaim: is the `storageclass.yaml` automatically resolved?

File: infrastructure/kube/keep-test/monitoring/README.md [hunk 30]
```diff
@@ -0,0 +1,162 @@
...
+#### Prometheus
+
+Configuration is stored in [./prometheus/kustomization.yaml](kustomization.yaml) file.
+
+To preview generated config run `kubectl kustomize prometheus/`
+
+To apply the configuration run `kubectl apply -k prometheus/`
+
+
+`kubectl diff -k prometheus/`
```
Comment (pdyraga) [inline infrastructure/kube/keep-test/monitoring/README.md:162]: Looks like some orphaned info.

File: infrastructure/kube/keep-test/monitoring/grafana-service.yaml (hunk 10 above)
Comment (pdyraga) [reply infrastructure/kube/keep-test/monitoring/grafana-service.yaml]: Did we figure it out?

File: infrastructure/kube/keep-test/monitoring/README.md [hunk 31]
```diff
@@ -0,0 +1,162 @@
...
+For keep-client monitoring purpose `client-dashboards.json` holds template config for that service
+
+One ConfigMap YAML file contains two section of configmaps configuration for two different services. 
+
+#### III. Create Prometheus PersistentVolumeClaim
+
+Create a PVC object
+
+  ```bash
+  kubectl create -f prometheus-pvc.yaml
```
[the remark this answers is not in the document]
Comment (nkuba) [reply infrastructure/kube/keep-test/monitoring/README.md]: 80d1e918fd70af7a674076c68cc3d1ded6aca5da

File: infrastructure/kube/keep-test/monitoring/prometheus/prometheus-deployment.yaml [hunk 32]
```diff
@@ -0,0 +1,72 @@
...
+      type: monitoring
+  template:
+    spec:
+      containers:
+        - name: prometheus
+          image: prom/prometheus:v2.37.0
+          args:
+            - --config.file=/etc/prometheus/config.yaml
+            - --storage.tsdb.path=/etc/prometheus/data
+            - --storage.tsdb.retention.time=1y
```
[the remark this answers is not in the document]
Comment (nkuba) [reply infrastructure/kube/keep-test/monitoring/prometheus/prometheus-deployment.yaml:20]: Of course! 
> Prometheus stores an average of only 1-2 bytes per sample. Thus, to plan the capacity of a Prometheus server, you can use the rough formula:
>
> needed_disk_space = retention_time_seconds * ingested_samples_per_second * bytes_per_sample

[source](https://prometheus.io/docs/prometheus/latest/storage/)
😉 

So right now our keep-test Prometheus instance stores a lot of data from Kubernetes cluster that we're not using but could be used to set up dashboards. For production, we will collect just the Keep Nodes data, so the disk storage will be reduced.

File: infrastructure/kube/keep-test/monitoring/monitoring-ingress.yaml (hunk 27 above)
Comment (nkuba) [reply infrastructure/kube/keep-test/monitoring/monitoring-ingress.yaml:3]: 2a0f8f385af5dac04897c37ca6377ddeacab1fbb

**Comment by nkuba** [comment]
Closed in favor of https://github.com/keep-network/keep-core/pull/3353.

**Pull request closed without merging**